### PR TITLE
Well-orderings and Ordinals

### PR DESCRIPTION
**Pull request by LuuBluum**
Defines well-ordered sets, their various properties, as well as a definition of ordinals and some ordinal arithmetic (namely, successor, addition, and multiplication; exponentiation will have to wait).

### CODE REVIEW

**Comment by felixwellen** [comment]
Please rebase/merge master

File: Cubical/Relation/Binary/Order/Woset/Base.agda [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+private
+  variable
+    ℓ ℓ' ℓ'' ℓ₀ ℓ₀' ℓ₁ ℓ₁' ℓ₂ ℓ₂' : Level
+
+record IsWoset {A : Type ℓ} (_≺_ : A → A → Type ℓ') : Type (ℓ-max ℓ ℓ') where
+  no-eta-equality
+  constructor iswoset
+
+  field
+    is-set : isSet A
```
Comment (felixwellen) [inline Cubical/Relation/Binary/Order/Woset/Base.agda:43]: This follows from prop-valued and weak extensionality.

File: Cubical/Relation/Binary/Order/Woset/Simulation.agda [hunk 2]
```diff
@@ -0,0 +1,926 @@
...
+open import Cubical.Relation.Binary.Base
+open import Cubical.Relation.Binary.Extensionality
+open import Cubical.Relation.Binary.Order.Properties
+
+private
+  variable
+    ℓ ℓ' ℓa ℓa' ℓb ℓb' ℓc ℓc' ℓd ℓd' : Level
+
+-- We start with the truncated version and derive the untruncated
+isSimulation∃ : (A : Woset ℓa ℓa') (B : Woset ℓb ℓb') (f : ⟨ A ⟩ → ⟨ B ⟩)
```
Comment (felixwellen) [inline Cubical/Relation/Binary/Order/Woset/Simulation.agda:31]: It would be nice to extend this comment a bit (the truncated version of what?).

**Review by felixwellen — CHANGES_REQUESTED** [review]
Sorry for the long wait (it does help if PRs are shorter)
and thanks for contributing!
Apart from some details which I commented on, I do not know how your PR relates to the literature. @iblech kindly pointed me to this article (not the only one on the topic):
https://arxiv.org/pdf/2301.10696.pdf
which also comes with an agda formalization and an equivalent definition of ordinal. You should add pointers to the literature and comment on the relation of your formalization and others (e.g. if it is independet/inspired/partially copied).

**Comment by LuuBluum** [comment]
> Sorry for the long wait (it does help if PRs are shorter) and thanks for contributing! Apart from some details which I commented on, I do not know how your PR relates to the literature. @iblech kindly pointed me to this article (not the only one on the topic): https://arxiv.org/pdf/2301.10696.pdf which also comes with an agda formalization and an equivalent definition of ordinal. You should add pointers to the literature and comment on the relation of your formalization and others (e.g. if it is independet/inspired/partially copied).

Oh, to explain that, this is identical to the formalization done in the HoTT book, outside of the bit at the end for ordinal suprema which I did cite my source.

**Comment by felixwellen** [comment]
Good - that would be great!

File: Cubical/Relation/Binary/Order/Woset/Base.agda (hunk 1 above)
Comment (LuuBluum) [reply Cubical/Relation/Binary/Order/Woset/Base.agda:43]: It does, but it helps to keep all of these orderings consistent and require the basics that they're prop-valued and over sets. For instance, posets require that A is a set, but that follows from prop-valued, reflexive, and antisymmetry.

File: Cubical/Relation/Binary/Order/Woset/Simulation.agda [hunk 3]
```diff
@@ -0,0 +1,926 @@
...
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Function
+open import Cubical.Foundations.HLevels
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Structure
+
+open import Cubical.HITs.PropositionalTruncation as ∥₁
+open import Cubical.HITs.SetQuotients as /₂
```
[the remark this answers is not in the document]
Comment (LuuBluum) [reply Cubical/Relation/Binary/Order/Woset/Simulation.agda:12]: To be completely fair, when it comes to shorthand import naming in the library, there's not much of common practice anywhere. I've seen anything from PT to PropTrunc to PropositionalTruncation to just Prop (which seems especially dubious). I went with the notation that fits best with the idea of using the operator symbol for the import (similar to using the sum operator for importing sum, or N for naturals, and so on).

That and, by extension to the fact that I had already been doing this, but across all of the binary relation subfolder, propositional truncation is imported as `∥₁`. So, technically speaking, it's more consistent with the rest of the contents of these folders.

[the remark this answers is not in the document]
Comment (felixwellen) [reply Cubical/Relation/Binary/Order/Woset/Simulation.agda:12]: Yeah ;-) 
With "common practice" I mean, that some people agreed to do it like this once, but I am also fine with local name consistency.

**Comment by felixwellen** [comment]
Looks all good to me now! -> merging.

**Review by felixwellen — APPROVED** [review]

**Pull request merged**